DR-2494: Add DRS post endpoints to support RAS passport Auth

snf2ye opened the pull request:
Note to reviewers:
- I broke down the changes into meaningful commits.  Otherwise, the changes associated with a relatively small change (adding a directory "auth" to store "iam" and "ras") overwhelmed the PR. 
- I've added comments for areas that I'd particularly like feedback on
- I'm least confident about how I wired in the ECMService, so I would definitely appreciate feedback there

### Change Log
- Utilize the External Credentials Manager (ECM) to validate passports. This means we're working with a new service. Follow-on PRs will be needed to reference the ECM URL for other environments (alpha, staging, prod, etc.) to our helm definitions repo.
- Add two new endpoints: PostObjectId and PostAccessURL
- Attempt to gracefully share the bulk of the DRSService code between GetObjectId & PostObjectID and then between GetAccessURL and PostAccessURL.
- Added parent "auth" folder to store IAM and ECM code. Definitely would appreciate feedback on whether this makes sense - I wasn't sure where else to put the ECM service code. For the subfolders - Would a "ecm" folder make more sense than a "ras" folder?
- Added unit tests that should sufficiently confirm that we are throwing an exception when the user provides an invalid passport or when the passport does not align with snapshot's PHSID/consent code.
- Add audit logging for response from ecm validate passport call
- **Potentially Breaking Change:** For the PostObject() endpoint, I'm only returning an access_id, NOT an access_url. I've changed the behavior for Azure to match this behavior. I've reached out on slack to users that this might potentially impact and am awaiting feedback. So far, the batch team seems okay with this change. 

**Other add-on changes:**
- We referenced "region.getValue()" and "region.getString()" seemingly interchangeably. As far as I can tell, they return the same value, so I've made it so we always use 

### Background
* More info on PHSIds and consent codes in [this doc](https://docs.google.com/document/d/10OW4Em_x88E8RxGASwc-tnlFbLtqdBZ5ewLowM49xYQ/edit).

The overarching idea of this PR is illustrated in the diagram below: We want the user to be able to provide a RAS passport and a DRS Object Id and get back a signed url.
![TDR File Access](https://user-images.githubusercontent.com/13254229/161851376-58bc9490-aa20-46a3-abac-16e7c19b6674.png)

Following the DRS spec, this is done in two steps:
![DRS PostObject Endpoint](https://user-images.githubusercontent.com/13254229/161852637-fd5a77e2-1fe0-4f25-8609-234c9a442341.png)

snf2ye committed:
Add "auth" folder containing iam and ecm; update imports

Fix test from re-org

snf2ye committed:
Add two new post DRS endpoints; Import ECM Service; Update DrsService to share code with new functions

Reorg to work with tests

Remove unneeded reference

snf2ye committed:
Add unit tests

reorg unit tests into smaller bites

Add full method unit tests for post object

add post access url endpoint tests

snf2ye commented on src/main/java/bio/terra/app/configuration/ApplicationConfiguration.java:
```diff
@@ -397,6 +399,11 @@ public ExecutorService performanceThreadpool() {
         new LinkedBlockingQueue<>(getMaxPerformanceThreadQueueSize()));
   }
 
+  @Bean
+  public RestTemplate restTemplate(RestTemplateBuilder builder) {
+    return builder.build();
+  }
+
```
I don't really understand why we seemingly need this RestTemplate bean. It appears we reference a rest template without the bean w/ some of our other referenced services.

nmalfroy replied:
How do you mean?  Was it complaining about not having a bean?  Could be that the ECM client requires it?  FWIW, not a big deal since I'm about to put up a PR that needs a Rest template bean :)

snf2ye replied:
Ok! I think actually referencing the RestTemplate bean as you suggested cleaned this up and I'm no longer concerned about it! Thank you!

snf2ye commented on src/main/java/bio/terra/service/auth/ras/ECMService.java:
```diff
@@ -0,0 +1,45 @@
+  private final ECMConfiguration ecmConfiguration;
+  private final MappingJackson2HttpMessageConverter jacksonConverter;
+
+  public ECMService(
+      ECMConfiguration ecmConfiguration, MappingJackson2HttpMessageConverter jacksonConverter) {
+    this.ecmConfiguration = ecmConfiguration;
+    this.jacksonConverter = jacksonConverter;
+  }
+
+  public PassportApi getPassportApi() {
+    // ECM returns lots of responses as JSON string.
+    // By default, these get intercepted by Spring's raw string converter, but they're quoted, so
+    // they should be parsed as JSON.
+    // N.B. This means that any non-JSON responses won't be parsed correctly, so if ECM ever returns
+    // any other content-type, this would have to be revised. However, that seems unlikely.
+    var restTemplate = new RestTemplate(List.of(jacksonConverter));
+
+    var client = new ApiClient(restTemplate);
+    client.setBasePath(ecmConfiguration.getInstanceUrl());
+
+    return new PassportApi(client);
+  }
```
This is all stolen from the DRSHub repo. Not sure if there is an easier way for us to reference this service.

nmalfroy replied:
I think this is reasonable.  It looks pretty much like how it does for the Sam client

snf2ye commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -146,69 +180,122 @@ public void close() {
+            .criteria(List.of(criteria));
+    var result = ecmService.validatePassport(request);
+
+    // log audit info
+    logger.info(result.getAuditInfo().toString());
```
@nmalfroy  - Is this sufficient for the audit logging?

nmalfroy replied:
I think we need may need to be more explicit.  If you look here https://docs.google.com/document/d/1isNIYTkoOM0TdQ3skcSu79VMh8CCPHq_Zh356VXSVDE/edit#heading=h.ab5n175zvcrt there are the fields that we need.  I'm wondering if a simple formatted message is enough here.

snf2ye replied:
Thanks for discussing offline. 
Example log:
`[Validate Passport Audit]: Data Repository accessed: TDR, Study/Data set accessed: <redacted>, Date/Time of access: 04-11-2022 16:02:40 EST, ECM Audit Info: {passport_jti=<redacted>, external_user_id=<redacted>, internal_user_id=<redacted>}`

nmalfroy commented on src/test/java/bio/terra/service/tabulardata/google/BigQueryPdaoTest.java:
```diff
@@ -25,11 +25,11 @@
 import bio.terra.model.IngestRequestModel;
 import bio.terra.model.SnapshotModel;
 import bio.terra.model.SnapshotSummaryModel;
+import bio.terra.service.auth.iam.IamProviderInterface;
```
FWIW, I agree with your naming.  Nesting in an `auth` package makes a ton of sense

nmalfroy commented on src/main/java/bio/terra/app/configuration/ECMConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+@Configuration
+@EnableConfigurationProperties
+@ConfigurationProperties(prefix = "ecm")
+public class ECMConfiguration {
```
Curious about your thoughts: should this be `ECM` or `Ecm`?  I always struggle with this.  I'd lean towards the later but I'm struggling to find other example of this

nmalfroy replied:
Eh, you know what, never mind. I see that we have DRS using all caps so let's keep that!  I'll leave my comment in case others go through the same thought

nmalfroy commented on src/main/java/bio/terra/app/configuration/ECMConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+
+  private String instanceUrl;
+
+  public String getInstanceUrl() {
```
I think we've ben typically using `basePath` to point to terra service URLs (e.g. Sam and Terra configs). I'd use the same here

nmalfroy commented on src/main/java/bio/terra/service/auth/ras/ECMService.java:
```diff
@@ -0,0 +1,45 @@
+    // they should be parsed as JSON.
+    // N.B. This means that any non-JSON responses won't be parsed correctly, so if ECM ever returns
+    // any other content-type, this would have to be revised. However, that seems unlikely.
+    var restTemplate = new RestTemplate(List.of(jacksonConverter));
```
Can you not use the rest template bean that you created above?

snf2ye replied:
Oh you're totally right - that cleans this up quite a bit.

nmalfroy commented on src/main/java/bio/terra/service/auth/ras/ECMService.java:
```diff
@@ -0,0 +1,45 @@
+  private final ECMConfiguration ecmConfiguration;
+  private final MappingJackson2HttpMessageConverter jacksonConverter;
+
+  public ECMService(
```
Do we need an `@Autowired`?

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -131,6 +139,32 @@ public void close() {
   }
 
+  /*
+  Get/Post DRS Object
```
The extra comment looks a little strange here.  Do we even need it?  Or maybe just a space to separate the comments

snf2ye replied:
Removed

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -131,6 +139,32 @@ public void close() {
+  public DRSObject lookupObjectByDrsIdPassport(
+      String drsObjectId, DRSPassportRequestModel drsPassportRequestModel) {
+    try (DrsRequestResource r = new DrsRequestResource()) {
+      Snapshot snapshot = lookupSnapshotForDRSObject(drsObjectId);
```
Given the optimization that Trevyn just had to do: we should use the cached object I think

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -146,69 +180,122 @@ public void close() {
-        throw new DrsObjectNotFoundException(
-            "No snapshot found for DRS object id '" + drsObjectId + "'", ex);
-      }
+      Snapshot snapshot = lookupSnapshotForDRSObject(drsObjectId);
```
Same comment about the cache

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -146,69 +180,122 @@ public void close() {
-        return drsObjectFromFSDir((FSDir) fsObject, drsId.getSnapshotId());
-      }
+  @VisibleForTesting
+  Snapshot lookupSnapshotForDRSObject(String drsObjectId) {
```
Given how big a snapshot is, I think we should probably avoid using the entire snapshot object if we can avoid it.  I think the `SnapshotCacheResult` object in #1253 might be worthwhile here

snf2ye replied:
Definitely! Once that PR goes in, I'll add the needed fields to the cache result.

snf2ye replied:
Should be added now!

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -146,69 +180,122 @@ public void close() {
+    }
+    // Pass the passport + phs id + consent code to ECM
+    var criteria = new RASv1Dot1VisaCriterion().consentCode(consentCode).phsId(phsId);
+    criteria.issuer("https://stsstg.nih.gov").type("RASv1Dot1VisaCriterion");
```
These should be in constants

snf2ye replied:
Good point - Are constants in the file good enough or should I pull them to app properties?

nmalfroy replied:
I think in the class is totally reasonable

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -220,41 +307,41 @@ public DRSAccessURL getAccessUrlForObjectId(
-  private Optional<DRSAccessMethod> getAccessMethodMatchingAccessId(
-      String accessId, DRSObject object, TypeEnum methodType) {
+  /*
+  Helper methods
```
Kind of a strange format for the log

snf2ye replied:
Removed

nmalfroy reviewed (changes requested):
This looks great.  Thank you for breaking up the commits like this, it made it a ton easier to review.  I did have a couple of comments to look into around perf (especially in light of some of the things we ran into yesterday)

samanehsan commented on src/main/resources/api/data-repository-openapi.yaml:
```diff
@@ -2862,6 +2862,75 @@ paths:
+          schema:
+            type: string
+      security:
+        - PassportAuth: [ ]
```
What is this `security` field used for? I haven't seen that before!

snf2ye replied:
Thanks for asking about this! I had copied it from the DRS Spec, but overlooked revisiting what it meant. 
- In general, here's more details on the security field in swagger - https://swagger.io/docs/specification/authentication/
- I'm adding a change to add a security schema to define "passportauth" --> This definition is also copied from the DRS spec. https://github.com/ga4gh/data-repository-service-schemas/blob/master/openapi/data_repository_service.openapi.yaml#L130

Basically, by defining the security for these endpoints, we're overriding [the oauth security descriptor for the rest of the endpoints ](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/resources/api/data-repository-openapi.yaml#L33)

snf2ye replied:
As I mentioned in standup, I decided to remove this security option. As planned, the user will just provide the passport as part of the request and we will make this endpoint unauthenticated.

samanehsan commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -297,34 +383,39 @@ private DRSAccessURL signGoogleUrl(Snapshot snapshot, DRSAccessMethod accessMeth
-      accessMethods = getDrsAccessMethodsOnGcp(snapshotId, fsFile, authUser);
+      String gcpRegion = retrieveGCPSnapshotRegion(snapshot, fsFile);
+      if (passportAuth) {
+        accessMethods = getDrsSignedURLAccessMethods("gcp-passport-", gcpRegion);
```
What do you think about making these prefixes constants?

snf2ye replied:
Good point! Should be constants now.

samanehsan commented on src/test/java/bio/terra/service/filedata/DrsServiceTest.java:
```diff
@@ -188,6 +213,145 @@ public void testLookupNegative() {
 
+  @Test
+  public void testLookupSnapshot() {
+    drsService.lookupSnapshotForDRSObject(googleDrsObjectId);
```
Can we assert that this retrieves the expected snapshot?

snf2ye replied:
Good idea - Added!

samanehsan reviewed (approved):
Once the comments from Muscles are addressed this LGTM! Thanks for structuring the commits so clearly! :octocat:

snf2ye committed:
Code Review Changes

* Code review: add security schema for passport auth
* code review: Switch to base path
* Code review: Remove comments
* Code review: Remove comments
* Code review: Move to constants
* Code review: Use new RestTemplate bean
* Format Log message
* Revert Code review change: I actually don't think we want this as an auth option since included in request
*spotbugs fix
* Code review: move prefixes to constants
* Code Review: Assert retrieving correct snapshot
* Remove transaction id from log since we no longer have it

rework log to just return audit log value

rework

Fix date for log

snf2ye committed:
Fix test

samanehsan reviewed (approved):
The new commits look good to me! :octocat:

snf2ye committed:
spotlessApply after handling merge conflicts

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -187,63 +193,73 @@ public DRSObject lookupObjectByDrsIdPassport(
-    // log audit info
-    logger.info(result.getAuditInfo().toString());
+    var df = new SimpleDateFormat("MM-dd-yyyy HH:mm:ss z");
+    df.setTimeZone(TimeZone.getTimeZone("EST"));
```
Do we want this to be EST?  I think we'd be better served making this UTC

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -75,6 +77,11 @@ public class DrsService {
   private final Logger logger = LoggerFactory.getLogger(DrsService.class);
 
+  private static final String ACCESS_ID_PREFIX_GCP = "gcp-";
+  private static final String ACCESS_ID_PREFIX_AZURE = "azure-";
```
I think this was `az-` originally.  Is it ok that it would change?  Should we leave it as `az-`?

nmalfroy commented on src/main/java/bio/terra/service/filedata/DrsService.java:
```diff
@@ -78,7 +78,7 @@ public class DrsService {
 
   private static final String ACCESS_ID_PREFIX_GCP = "gcp-";
-  private static final String ACCESS_ID_PREFIX_AZURE = "azure-";
+  private static final String ACCESS_ID_PREFIX_AZURE = "az-";
```
lol...that's what I get for going commit-by-commit :)

nmalfroy reviewed (approved):
Looks great, thanks for making the changes!  Just one thing really about the timezone

snf2ye committed:
Code review: Swap timezones
